frontend/cypress/integration/common/apps.ts

```diff
@@ -135,3 +135,16 @@ When('user {string} toggle {string}', function (action: 'checks' | 'unchecks', t
 Then('the {string} column {string}', (col: string, action: 'appears' | 'disappears') => {
   colExists(col, action === 'appears');
 });
+
+Then('user may only see {string}', (sees: string) => {
+  cy.get('tbody').within(() => {
+    cy.get('tr').should('have.length', 1);
+    cy.get('td').then(td => {
+      if (td.length === 1) {
+        cy.get('h5').contains('No applications found');
+      } else {
+        cy.contains('tr', sees);
```

Comment:
I'm not completely sure how this works.
Does `if else` condition handles the presence of traffic generator?

Reply:
Well this code is generic but yes, this handles the fact that in different CI environments we may or may not have the traffic generator or it may or may not have a sidecar injected.  It just allows for it to be there or not be there in the apps list.  It's a concession to our different needs in different places.

So now in the feature you see `Then user may only see "kiali-traffic-generator"` because in some CI (downstream) it's expected to be there and in some places it isn't (upstream).

Reply:
Thanks for explanation. :+1: 
Noting for myself - it will be interesting to see if we come across more exceptions like this in future